src/transpile/fusion.hpp

```diff
@@ -32,6 +32,645 @@ using oplist_t = std::vector<op_t>;
+      for (size_t i = 0; i < vec.size(); ++i)
+        vec[i] = fusioned_op.mats[0](i, i);
+      fusioned_op = Operations::make_diagonal(fusioned_op.qubits, vec, std::string("fusion"));
+    }
+
+    return fusioned_op;
+  };
+
+  virtual op_t generate_operation_internal(const std::vector<op_t>& fusioned_ops,
+                                           const reg_t &fusioned_qubits) const = 0;
+
+  virtual bool can_apply(const op_t& op, uint_t max_fused_qubits) const = 0;
+
+  virtual bool can_ignore(const op_t& op) const {
+    switch (op.type) {
+      case optype_t::barrier:
+        return true;
+      case optype_t::gate:
+        return op.name == "id" || op.name == "u0";
+      default:
+        return false;
+    }
+  }
+
+  static FusionMethod& find_method(const Circuit& circ,
+                                  const opset_t &allowed_opset,
+                                  const bool allow_superop,
+                                  const bool allow_kraus);
+
+  static bool exist_non_unitary(const std::vector<op_t>& fusioned_ops) {
+    for (auto & op: fusioned_ops)
+      if (noise_opset_.contains(op.type))
+        return true;
+    return false;
+  };
+
+private:
+  const static Operations::OpSet noise_opset_;
+};
+
+const Operations::OpSet FusionMethod::noise_opset_(
+  {Operations::OpType::kraus,
+   Operations::OpType::superop,
+   Operations::OpType::reset},
+  {}, {}
+);
+
+class UnitaryFusion : public FusionMethod {
+public:
+  virtual std::string name() override { return "unitary"; };
+
+  virtual bool support_diagonal() const override { return true; }
+
+  virtual op_t generate_operation_internal (const std::vector<op_t>& fusioned_ops,
+                                           const reg_t &qubits) const override {
+    // Run simulation
+    RngEngine dummy_rng;
+    ExperimentResult dummy_result;
+
+    // Unitary simulation
+    QubitUnitary::State<> unitary_simulator;
+    unitary_simulator.initialize_qreg(qubits.size());
+    unitary_simulator.apply_ops(fusioned_ops, dummy_result, dummy_rng);
+    return Operations::make_unitary(qubits, unitary_simulator.qreg().move_to_matrix(),
+                                    std::string("fusion"));
+  };
+
+  virtual bool can_apply(const op_t& op, uint_t max_fused_qubits) const {
+    if (op.conditional)
+      return false;
+    switch (op.type) {
+      case optype_t::matrix:
+        return op.mats.size() == 1 && op.qubits.size() <= max_fused_qubits;
+      case optype_t::diagonal_matrix:
+        return op.qubits.size() <= max_fused_qubits;
+      case optype_t::gate: {
+        if (op.qubits.size() > max_fused_qubits)
+          return false;
+        return QubitUnitary::StateOpSet.contains_gates(op.name);
+      }
+      default:
+        return false;
+    }
+  };
+};
+
+class SuperOpFusion : public UnitaryFusion {
+public:
+  virtual std::string name() override { return "superop"; };
+
+  virtual bool support_diagonal() const override { return false; }
```

Comment:
We could add a diagonal_superop instruction for the density matrix / superop simulators